### PR TITLE
Added unit tests for HabitFactSpecification

### LINKED ISSUE #11: Create unit test to HabitFactSpecification.java
###  Create Unit Tests for `HabitFactSpecification` (`GreenCityMVP`)

**File:**  
`service/src/main/java/greencity/filters/HabitFactSpecification.java`

### Description

Develop a set of unit tests for the `HabitFactSpecification` class, which builds JPA criteria-based queries for the `HabitFact` entity based on a list of `SearchCriteria`.

This specification supports filtering by:
- `id`
- `habitId`
- `content` (translated field via a join)

### Input Data

- `SearchCriteria` entries (type, value)
- Entity: `HabitFact`
- Joins with: `Habit`, `HabitFactTranslation`

###  Test Scenarios

- [x] `toPredicate(...)` returns a valid predicate when searching by `"id"`
- [x] `toPredicate(...)` handles filtering by `"habitId"` with a join
- [x] `toPredicate(...)` filters by `"content"` (join with `HabitFactTranslation`)
- [x] Empty string for `"content"` should not add additional predicates
- [x] Combination of multiple `SearchCriteria` types builds a compound predicate
- [x] Covers null or empty value edge case handling

###  Acceptance Criteria

- [x] All tests pass (`JUnit 5`)
- [x] `CriteriaBuilder`, `Root`, `CriteriaQuery`, and `Join` are properly mocked with `Mockito`
- [x] Each `if` block in `toPredicate(...)` is tested and covered
- [x] Private helper methods (`getHabitIdPredicate`, `getTranslationPredicate`) are tested indirectly via public method
- [x] **100% line and branch coverage** of `HabitFactSpecification`

 **Tip**:  
Use `@ExtendWith(MockitoExtension.class)` and mock `Root<HabitFact>`, `CriteriaBuilder`, `CriteriaQuery` explicitly.  
Focus on verifying that predicates are combined correctly and appropriate joins are created.

### PR DESCRIPTION
**Pull request by Shcherba04**
toPredicate Id, HabitIdwithJoin, Content, ContentNull testing, MultipleCriteria

## Summary by CodeRabbit

* **Tests**
  * Added comprehensive unit tests to improve filtering and search functionality for habits, ensuring accurate results for various search conditions.

### CODE REVIEW

File: service/src/test/java/greencity/filters/HabitFactSpecificationTest.java [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.Mockito.*;
+
+
+@ExtendWith(MockitoExtension.class)
+class HabitFactSpecificationTest {
+
+    @Mock
+    private Root<HabitFact> HabitFactRoot;
```
Comment (natri0) [inline service/src/test/java/greencity/filters/HabitFactSpecificationTest.java]: fields should be in camelCase like rabbit suggested
```suggestion
    @Mock
    private Root<HabitFact> habitFactRoot;
```

File: service/src/test/java/greencity/filters/HabitFactSpecificationTest.java [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+    @Mock
+    private CriteriaBuilder builder;
+
+    @Mock
+    private CriteriaQuery MockCriteriaQuery;
```
Comment (natri0) [inline service/src/test/java/greencity/filters/HabitFactSpecificationTest.java]: same as above
```suggestion
    @Mock
    private CriteriaQuery mockCriteriaQuery;
```

File: service/src/test/java/greencity/filters/HabitFactSpecificationTest.java [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    HabitFactSpecification specification;
+
+    @Test
+    void toPredicate_Id(){
+        List<SearchCriteria> searchCriteriaList = new ArrayList<>();
+        searchCriteriaList.add(SearchCriteria.builder()
+                .key("id")
+                .type("id")
+                .value(2L)
+                .build());
```
Comment (natri0) [inline service/src/test/java/greencity/filters/HabitFactSpecificationTest.java]: minor nitpick: use `List.of`
```suggestion
        List<SearchCriteria> searchCriteriaList = List.of(
                SearchCriteria.builder()
                        .key("id")
                        .type("id")
                        .value(2L)
                        .build());
```

File: service/src/test/java/greencity/filters/HabitFactSpecificationTest.java [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+    @Test
+    @MockitoSettings(strictness = Strictness.LENIENT)
+    void toPredicate_HabitIdWithJoin() {
+        List<SearchCriteria> searchCriteriaList = new ArrayList<>();
+        searchCriteriaList.add(SearchCriteria.builder()
+                .key("habitId")
+                .type("habitId")
+                .value(5L)
+                .build());
```
Comment (natri0) [inline service/src/test/java/greencity/filters/HabitFactSpecificationTest.java]: same as above, use List.of

File: service/src/test/java/greencity/filters/HabitFactSpecificationTest.java [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    }
+
+    @Test
+    void toPredicate_WithTranslationJoin(){
+        List<SearchCriteria> searchCriteriaList = new ArrayList<>();
+        searchCriteriaList.add(SearchCriteria.builder()
+                .key("content")
+                .type("content")
+                .value("Smth")
+                .build());
```
Comment (natri0) [inline service/src/test/java/greencity/filters/HabitFactSpecificationTest.java]: same as above, use List.of

File: service/src/test/java/greencity/filters/HabitFactSpecificationTest.java [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+    }
+
+    @Test
+    void toPredicate_ContentNull(){
+        List<SearchCriteria> searchCriteriaList = new ArrayList<>();
+        searchCriteriaList.add(SearchCriteria.builder()
+                .key("content")
+                .type("content")
+                .value(" ")
+                .build());
```
Comment (natri0) [inline service/src/test/java/greencity/filters/HabitFactSpecificationTest.java]: same as above, use List.of

File: service/src/test/java/greencity/filters/HabitFactSpecificationTest.java [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+        searchCriteriaList.add(SearchCriteria.builder()
+                .key("habitId")
+                .type("habitId")
+                .value(5L)
+                .build());
+        searchCriteriaList.add(SearchCriteria.builder()
+                .key("content")
+                .type("content")
+                .value("Smth")
+                .build());
```
Comment (natri0) [inline service/src/test/java/greencity/filters/HabitFactSpecificationTest.java]: same as above, use List.of

File: service/src/test/java/greencity/filters/HabitFactSpecificationTest.java [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+
+    @Mock
+    private Path<String> contentPath;
+
+    @Mock
+    private Predicate likePredicate;
+
+    @Mock
+    private Expression<String> as;
+    HabitFactSpecification specification;
```
Comment (natri0) [inline service/src/test/java/greencity/filters/HabitFactSpecificationTest.java]: for some predicates/paths/etc you use local variables, but for some others you use fields; maybe use local variables for everything, or at least document which fields correspond to which test?

**Review by natri0 — CHANGES_REQUESTED** [review]
some minor nitpicks, but is good otherwise.

File: service/src/test/java/greencity/filters/HabitFactSpecificationTest.java [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+                .key("habitId")
+                .type("habitId")
+                .value(5L)
+                .build());
+        searchCriteriaList.add(SearchCriteria.builder()
+                .key("content")
+                .type("content")
+                .value("Smth")
+                .build());
+        specification=new HabitFactSpecification(searchCriteriaList);
```
Comment (Kostya-Pav) [inline service/src/test/java/greencity/filters/HabitFactSpecificationTest.java]: You already use this logic multiple times. It’s better to move the shared setup into a @BeforeAll method to avoid duplication when preparing test objects.

File: service/src/test/java/greencity/filters/HabitFactSpecificationTest.java (hunk 8 above)
Comment (Shcherba04) [reply service/src/test/java/greencity/filters/HabitFactSpecificationTest.java]: Due lack of time, I can only document it.

File: service/src/test/java/greencity/filters/HabitFactSpecificationTest.java (hunk 9 above)
Comment (Shcherba04) [reply service/src/test/java/greencity/filters/HabitFactSpecificationTest.java]: If I use @BeforeAll or @BeforeEach, code need full rewrite. And I preparing to test for master degree, which would be on 08.07. So I can't implement this setup:(

**Review by natri0 — APPROVED** [review]

**Review by oleksandrYeriemieiev — APPROVED** [review]

File: service/src/test/java/greencity/filters/HabitFactSpecificationTest.java [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+    @MockitoSettings(strictness = Strictness.LENIENT)
+    void toPredicate_HabitIdWithJoin() {
+        List<SearchCriteria> searchCriteriaList = List.of(
+        SearchCriteria.builder()
+                .key("habitId")
+                .type("habitId")
+                .value(5L)
+                .build());
+        specification = new HabitFactSpecification(searchCriteriaList);
+        when(HabitFactRoot.join(HabitFact_.habit)).thenReturn(habitJoin);
```
Comment (khomenkooleksandr) [inline service/src/test/java/greencity/filters/HabitFactSpecificationTest.java]: HaBitFactRoot  code convention habitRactRoot

File: service/src/test/java/greencity/filters/HabitFactSpecificationTest.java [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+    private Join<HabitFact,Habit> habitJoin;
+    
+    // For content tests 
+    
+    @Mock
+    private Path<String> contentPath;
+
+    //Translation join 
+    @Mock
+    private Predicate likePredicate;
```
Comment (khomenkooleksandr) [inline service/src/test/java/greencity/filters/HabitFactSpecificationTest.java:53]: variable is not used

File: service/src/test/java/greencity/filters/HabitFactSpecificationTest.java [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+
+    @Mock
+    private Expression<String> as;
```
Comment (khomenkooleksandr) [inline service/src/test/java/greencity/filters/HabitFactSpecificationTest.java]: variable is not use

File: service/src/test/java/greencity/filters/HabitFactSpecificationTest.java [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+                .type("habitId")
+                .value(5L)
+                .build());
+        specification = new HabitFactSpecification(searchCriteriaList);
+        when(HabitFactRoot.join(HabitFact_.habit)).thenReturn(habitJoin);
+        when(habitJoin.get(Habit_.id)).thenReturn(habitIdPath);
+        Predicate mockPredicate = mock(Predicate.class);
+        when(builder.equal(habitIdPath, 5L)).thenReturn(mockPredicate);
+        Predicate conjunctionPredicate = mock(Predicate.class);
+        when(builder.conjunction()).thenReturn(conjunctionPredicate);
```
Comment (khomenkooleksandr) [inline service/src/test/java/greencity/filters/HabitFactSpecificationTest.java:107]: Code Duplication:
The test cases repeat similar setup logic for mocking builder.conjunction(), builder.and(), etc. This could be refactored into a common setup method  @BeforeEach block to reduce duplication.

File: service/src/test/java/greencity/filters/HabitFactSpecificationTest.java [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+        SearchCriteria.builder()
+                .key("habitId")
+                .type("habitId")
+                .value(5L)
+                .build(),
+        SearchCriteria.builder()
+                .key("content")
+                .type("content")
+                .value("Smth")
+                .build());
```
Comment (khomenkooleksandr) [inline service/src/test/java/greencity/filters/HabitFactSpecificationTest.java]: Hardcoded Values:
The test data ( 2L for id, 5L for habitId, "Smth" for content etc) is hardcoded. Using constants

File: service/src/test/java/greencity/filters/HabitFactSpecificationTest.java [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+        verify(builder).equal(habitIdPath, 5L);
+        verify(builder).like(contentPath, "%Smth%");
+        verify(builder).equal(habitIdPath, habitFactRootIdPath);
+        verify(builder).and(conjunctionPredicate, idEqualPredicate);
+        verify(builder).and(firstCombination, habitIdEqualPredicate);
+        verify(builder).and(secondCombination, contentAndPredicate);
+
+        }
+
+
```
Comment (khomenkooleksandr) [inline service/src/test/java/greencity/filters/HabitFactSpecificationTest.java]: extra spaces

File: service/src/test/java/greencity/filters/HabitFactSpecificationTest.java [hunk 16]
```diff
@@ -0,0 +1,241 @@
...
+        when(habitJoin.get(Habit_.id)).thenReturn(habitIdPath);
+        when(builder.equal(habitIdPath, 5L)).thenReturn(habitIdEqualPredicate);
+
+        Root<HabitFactTranslation> habitFactTranslationRoot = mock(Root.class);
+        when(mockCriteriaQuery.from(HabitFactTranslation.class)).thenReturn(habitFactTranslationRoot);
+        when(habitFactTranslationRoot.get(HabitFactTranslation_.content)).thenReturn(contentPath);
+        when(habitFactTranslationRoot.get(HabitFactTranslation_.habitFact).get(HabitFact_.id)).thenReturn(habitIdPath);
+        when(HabitFactRoot.get(HabitFact_.id)).thenReturn(habitFactRootIdPath);
+
+        Predicate likePredicate = mock(Predicate.class);
```
Comment (khomenkooleksandr) [inline service/src/test/java/greencity/filters/HabitFactSpecificationTest.java]: why don't you use an already declared variable?

File: service/src/test/java/greencity/filters/HabitFactSpecificationTest.java [hunk 17]
```diff
@@ -0,0 +1,241 @@
...
+        when(builder.and(likePredicate, equalPredicate)).thenReturn(andPredicate);
+        when(builder.and(conjunctionPredicate, andPredicate)).thenReturn(andPredicate);
+
+        Predicate result = specification.toPredicate(HabitFactRoot, mockCriteriaQuery, builder);
+        assertNotNull(result);
+        assertEquals(andPredicate, result);
+    }
+
+    @Test
+    void toPredicate_ContentNull(){
```
Comment (khomenkooleksandr) [inline service/src/test/java/greencity/filters/HabitFactSpecificationTest.java]: toPredicateContentEmpty

**Review by khomenkooleksandr — CHANGES_REQUESTED** [review]
The test class is robust and covers the main functionality of HabitFactSpecification. By addressing the suggestions above—simplifying test logic, improving naming, adding edge cases, and reducing redundancy—you can make the tests more maintainable and easier to understand.
Missing Edge Cases:
The tests do not cover invalid or null SearchCriteria values (except for the empty content case). 
You checked only in method toPredicate_ContentNull empty content.
Adding tests for these edge cases would improve robustness.

File: service/src/test/java/greencity/filters/HabitFactSpecificationTest.java (hunk 13 above)
Comment (Shcherba04) [reply service/src/test/java/greencity/filters/HabitFactSpecificationTest.java:107]: To your comment, HabitFactSpecification id and habitid doesn't cover null values. So I added tests and they will have errors due that.
